Support orderer tls

inzamam-iqbal opened the pull request:
- change to nephos_config.yaml:

contains new object call `tls` which has two attribute `enable` (true or false)and `tls_ca` : points to a  tls ca if left blank uses cryptogen. 
```yaml
cas:
  ca-tls:
    namespace: cas-tls
    tls_cert: ./examples/ca-tls-nephos-local.pem
ordering:
  secret_genesis: hlf--genesis
  tls:
    enable: true
    tls_ca: ca-tls
```
- If only enable is set then nephos assumes user is using cryptogen to generate the TLS certs, if tls_ca is defined, neohos will deploy a fabric CA to act as TLS CA and will use that CA to generate all the TLS certs.
- Then nephos get the `tls.crt` and `tls.key` and put them in the same secret and save it as a secret which get mounted to `/var/hyperledger/tls/server/pair` in orderer pod and will be used for tls communication. 
- Then nephos will look for every orderer which has tls enabled and get their tls_ca_root cert and save it on each peer node as a secret which will get use to verify tls cert during communication with orederer.

inzamam-iqbal commented:
@nicolapaoli @alexvicegrab not perfect, didn't make any changes to unit tests. Before doing them I wanted to get your feedbacks. Also if the path I'm going is fine as the next step I can let the user define a CA for tls which will get hosted and used to generate tls certificates

alexvicegrab commented on nephos/fabric/utils.py:
```diff
@@ -49,6 +49,12 @@ def credentials_secret(secret_name, namespace, username, password=None):
     return secret_data
 
 
+def tls_secret(secret_name, namespace, keys_files_path):
```
Why make an extra function when we can just use secret_from_files directly?

alexvicegrab commented on nephos/fabric/crypto.py:
```diff
@@ -326,6 +330,23 @@ def cacerts_to_secrets(namespace, msp_path, user):
+
+    keys_files_path = {"cacert.pem": f"{tls_path}/ca.crt"}
+    secret_name = f"hlf--orderer-tlsrootcert"
+    tls_secret(secret_name=secret_name, namespace=namespace, keys_files_path=keys_files_path)
```
This should just use `secret_from_files`.

alexvicegrab commented on nephos/fabric/crypto.py:
```diff
@@ -326,6 +330,23 @@ def cacerts_to_secrets(namespace, msp_path, user):
         item_to_secret(namespace, msp_path, user, item)
 
 
+def tls_to_secrets(namespace, tls_path, username):
```
This is missing a DocString, otherwise we won't generate Sphinx documentation.

alexvicegrab commented on nephos/fabric/crypto.py:
```diff
@@ -326,6 +330,23 @@ def cacerts_to_secrets(namespace, msp_path, user):
+    tls_secret(secret_name=secret_name, namespace=namespace, keys_files_path=keys_files_path)
+
+
+def setup_tls(opts, msp_name, release, id_type):
```
This is also missing a DocString

alexvicegrab commented on nephos/fabric/utils.py:
```diff
@@ -105,6 +111,38 @@ def get_helm_pod(namespace, release, app, item=0):
     return get_pod(namespace, identifier, item=item)
 
 
+def get_org_tls_ca_cert(opts, msp_namespace):
```
Missing DocString

alexvicegrab commented on nephos/fabric/utils.py:
```diff
@@ -105,6 +111,38 @@ def get_helm_pod(namespace, release, app, item=0):
+            )
+
+
+def get_tls_path(opts, id_type, namespace, release):
```
Missing DocString

alexvicegrab commented on nephos/fabric/utils.py:
```diff
@@ -105,6 +111,38 @@ def get_helm_pod(namespace, release, app, item=0):
+        )
+
+
+def is_orderer_tls_true(opts):
```
Missing DocString

alexvicegrab commented on nephos/helpers/k8s.py:
```diff
@@ -246,6 +246,25 @@ def secret_read(name, namespace="default"):
     return secret.data
 
 
+def secret_from_files(secret, namespace, keys_files_path):
```
Can we not reuse much functionality from below function rather than copy-pasting most of the code?

alexvicegrab replied:
This comment still stands, we end up with two functions that are almost identical. I'd rather keep yours, @inzamam-iqbal, and remove the `secret_from_file` in favour of `secrets_from_files`

alexvicegrab commented on examples/dev/helm_values/AlphaMSP/hlf-ord/ord1.yaml:
```diff
@@ -8,11 +8,15 @@ persistence:
   type: Solo
   mspID: AlphaMSP
-
+  tls:
```
I would create a separate dev_tls example with these changes, and leave the basic dev example as it was.

inzamam-iqbal commented:
@alexvicegrab thank you for the feedback. I still didn't do code cleanup. I will create a separate dev-tls example as you suggested. Is the nephos_config.yaml format for tls fine? if so i will go ahead with the suggested changes. Then I will create a new qa-tls example where we get the certificates from a separate CA

alexvicegrab commented:
In principle, yes, I might even remove the `enable: true` flag, as if we populate the TLS-CA flag, we would expect to use TLS. Otherwise the direction and approach look fine to me.

alexvicegrab commented:
And yes, good idea to create the QA TLS  (and eventually Prod TLS) example also, once the development one is reliably working. Keep up the good work :)

inzamam-iqbal commented:
@alexvicegrab I updated the PR with docstrings for all methods, removed the tls_secret method and created seperate dev_tls example. 
My thought behind defining tls_enable in nephos config is that a user can set only tls_enable flag true without specifying tls_ca in that case we will use the materials created by cryptogen to setup tls without using a tls_ca
```yaml
ordering:
  secret_genesis: hlf--genesis
  tls:
    enable: true
```

alexvicegrab commented:
Noted, in that case, this should work OK.

By the way, looking through the Nephos config, I noticed this line:
https://github.com/hyperledger-labs/nephos/blob/master/examples/dev/nephos_config.yaml#L21

I think for cryptogen examples, we should keep the `ca` field of MSPs empty or Null.

inzamam-iqbal commented:
@alexvicegrab yah sure I will make it empty

alexvicegrab commented:
@inzamam-iqbal 

You have 3 commits incorrectly signed off. To fix, head to your local branch and run:

    git rebase HEAD~18 --signoff

Now your commits will have your sign off. Next run

    git push --force-with-lease origin support_orderer_tls

inzamam-iqbal commented:
@alexvicegrab I have the PR and the PR description. With this PR nephos supports having a seperate TLS CA. I have created a qa-tls example too.

inzamam-iqbal commented:
@alexvicegrab please restart the CI.

alexvicegrab commented on nephos/fabric/crypto.py:
```diff
@@ -102,14 +106,16 @@ def register_id(ca_namespace, ca, username, password, node_type="client", admin=
 
 
-def enroll_id(opts, ca, username, password):
+def enroll_id(opts, ca, username, password, cert_type="MSP", extra=""):
```
Nice, I like this @inzamam-iqbal

inzamam-iqbal replied:
Thank you

alexvicegrab commented on nephos/fabric/crypto.py:
```diff
@@ -326,6 +332,70 @@ def cacerts_to_secrets(namespace, msp_path, user):
 
 
+def tls_to_secrets(namespace, tls_path, username):
+    """Get the tls server.crt and server.key and create the relevant secret
```
@inzamam-iqbal, please capitalise "TLS" in comments, since it corresponds to Transport Layer Security

alexvicegrab commented on nephos/fabric/crypto.py:
```diff
@@ -326,6 +332,70 @@ def cacerts_to_secrets(namespace, msp_path, user):
+
+
+def setup_tls(opts, msp_name, release, id_type):
+    """Setup tls by saving ID to K8S secrets.
```
And here "tls" -> "TLS". Naturally not python function names, but comments and DocStrings should be tidy. Thanks, @inzamam-iqbal

alexvicegrab commented on nephos/fabric/utils.py:
```diff
@@ -105,6 +105,83 @@ def get_helm_pod(namespace, release, app, item=0):
+
+
+def get_tls_path(opts, id_type, namespace, release):
+    """Get path to the directory containing tls materials for a node
```
"tls" -> "TLS"

alexvicegrab commented on nephos/fabric/utils.py:
```diff
@@ -105,6 +105,83 @@ def get_helm_pod(namespace, release, app, item=0):
+            release (str): Name of release/node.
+
+        Returns:
+            path: path to the directory containing materials of a node
```
"str: Path ..."

Return arguments should declare the type of the argument, rather than its name. See:

https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

alexvicegrab commented on nephos/fabric/utils.py:
```diff
@@ -105,6 +105,83 @@ def get_helm_pod(namespace, release, app, item=0):
+            opts (dict): Nephos options dict.
+
+        Returns:
+            Boolean: return true if tls is enabled for orderer and false otherwise
```
Boolean in python is "bool" so:

"bool: Return true if TLS is enabled for orderer and false otherwise"

inzamam-iqbal commented:
@alexvicegrab In the existing `secret_from_file` we get the user input when filename is not mentioned. But that functionality is only used in composer. So I think I don't need to add that functionality to the new `secret_from_files` when we get rid of `secret_from_file`

alexvicegrab commented:
Good question, for composer code, we could transfer the asking for input out of the function since we do it once only, or add that as an input to click instead

inzamam-iqbal commented:
@alexvicegrab I have updated the PR. I have replaced the secret_from_file function.

alexvicegrab commented on nephos/fabric/utils.py:
```diff
@@ -17,7 +17,7 @@
 from kubernetes.client.rest import ApiException
 
-from nephos.helpers.k8s import Executer, secret_create, secret_from_file, secret_read
+from nephos.helpers.k8s import Executer, secret_create, secret_read, secret_from_files
```
Just a comment, I usually put these in alphabetic order, but you can correct in next PR.

alexvicegrab commented on nephos/composer/install.py:
```diff
@@ -31,7 +31,7 @@
     cm_read,
     ingress_read,
-    secret_create
+    secret_create,
```
Good attention to detail, @inzamam-iqbal. It's indeed better to end with comma to improve change detection for `git`, etc.

alexvicegrab commented:
@inzamam-iqbal, very happy to approve this PR. Good job!

alexvicegrab commented:
Just realised, we need to bump the version of Nephos in `setup.py` and in `docs/conf.py`, since these are quite extensive changes. Otherwise all good!

alexvicegrab commented:
Thanks @inzamam-iqbal, LGTM

alexvicegrab commented:
Restarting CI/CD
